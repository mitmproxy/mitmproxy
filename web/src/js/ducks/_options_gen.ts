/** Auto-generated by test_app.py:test_generate_options_js */
export interface OptionsState {
    add_upstream_certs_to_client_chain: boolean
    allow_hosts: string[]
    anticache: boolean
    anticomp: boolean
    block_global: boolean
    block_list: string[]
    block_private: boolean
    body_size_limit: string | undefined
    cert_passphrase: string | undefined
    certs: string[]
    ciphers_client: string | undefined
    ciphers_server: string | undefined
    client_certs: string | undefined
    client_replay: string[]
    client_replay_concurrency: number
    command_history: boolean
    confdir: string
    connect_addr: string | undefined
    connection_strategy: string
    console_focus_follow: boolean
    content_view_lines_cutoff: number
    dns_listen_host: string
    dns_listen_port: number
    dns_mode: string
    dns_server: boolean
    export_preserve_original_ip: boolean
    http2: boolean
    http2_ping_keepalive: number
    ignore_hosts: string[]
    intercept: string | undefined
    intercept_active: boolean
    keep_host_header: boolean
    key_size: number
    listen_host: string
    listen_port: number
    map_local: string[]
    map_remote: string[]
    mode: string
    modify_body: string[]
    modify_headers: string[]
    normalize_outbound_headers: boolean
    onboarding: boolean
    onboarding_host: string
    onboarding_port: number
    proxy_debug: boolean
    proxyauth: string | undefined
    rawtcp: boolean
    readfile_filter: string | undefined
    rfile: string | undefined
    save_stream_file: string | undefined
    save_stream_filter: string | undefined
    scripts: string[]
    server: boolean
    server_replay: string[]
    server_replay_ignore_content: boolean
    server_replay_ignore_host: boolean
    server_replay_ignore_params: string[]
    server_replay_ignore_payload_params: string[]
    server_replay_ignore_port: boolean
    server_replay_kill_extra: boolean
    server_replay_nopop: boolean
    server_replay_refresh: boolean
    server_replay_use_headers: string[]
    showhost: boolean
    ssl_insecure: boolean
    ssl_verify_upstream_trusted_ca: string | undefined
    ssl_verify_upstream_trusted_confdir: string | undefined
    stickyauth: string | undefined
    stickycookie: string | undefined
    stream_large_bodies: string | undefined
    tcp_hosts: string[]
    termlog_verbosity: string
    tls_version_client_max: string
    tls_version_client_min: string
    tls_version_server_max: string
    tls_version_server_min: string
    upstream_auth: string | undefined
    upstream_cert: boolean
    validate_inbound_headers: boolean
    view_filter: string | undefined
    view_order: string
    view_order_reversed: boolean
    web_columns: string[]
    web_debug: boolean
    web_host: string
    web_open_browser: boolean
    web_port: number
    web_static_viewer: string | undefined
    websocket: boolean
}

export type Option = keyof OptionsState

export const defaultState: OptionsState = {
    add_upstream_certs_to_client_chain: false,
    allow_hosts: [],
    anticache: false,
    anticomp: false,
    block_global: true,
    block_list: [],
    block_private: false,
    body_size_limit: undefined,
    cert_passphrase: undefined,
    certs: [],
    ciphers_client: undefined,
    ciphers_server: undefined,
    client_certs: undefined,
    client_replay: [],
    client_replay_concurrency: 1,
    command_history: true,
    confdir: "~/.mitmproxy",
    connect_addr: undefined,
    connection_strategy: "eager",
    console_focus_follow: false,
    content_view_lines_cutoff: 512,
    dns_listen_host: "",
    dns_listen_port: 53,
    dns_mode: "regular",
    dns_server: false,
    export_preserve_original_ip: false,
    http2: true,
    http2_ping_keepalive: 58,
    ignore_hosts: [],
    intercept: undefined,
    intercept_active: false,
    keep_host_header: false,
    key_size: 2048,
    listen_host: "",
    listen_port: 8080,
    map_local: [],
    map_remote: [],
    mode: "regular",
    modify_body: [],
    modify_headers: [],
    normalize_outbound_headers: true,
    onboarding: true,
    onboarding_host: "mitm.it",
    onboarding_port: 80,
    proxy_debug: false,
    proxyauth: undefined,
    rawtcp: true,
    readfile_filter: undefined,
    rfile: undefined,
    save_stream_file: undefined,
    save_stream_filter: undefined,
    scripts: [],
    server: true,
    server_replay: [],
    server_replay_ignore_content: false,
    server_replay_ignore_host: false,
    server_replay_ignore_params: [],
    server_replay_ignore_payload_params: [],
    server_replay_ignore_port: false,
    server_replay_kill_extra: false,
    server_replay_nopop: false,
    server_replay_refresh: true,
    server_replay_use_headers: [],
    showhost: false,
    ssl_insecure: false,
    ssl_verify_upstream_trusted_ca: undefined,
    ssl_verify_upstream_trusted_confdir: undefined,
    stickyauth: undefined,
    stickycookie: undefined,
    stream_large_bodies: undefined,
    tcp_hosts: [],
    termlog_verbosity: "info",
    tls_version_client_max: "UNBOUNDED",
    tls_version_client_min: "TLS1_2",
    tls_version_server_max: "UNBOUNDED",
    tls_version_server_min: "TLS1_2",
    upstream_auth: undefined,
    upstream_cert: true,
    validate_inbound_headers: true,
    view_filter: undefined,
    view_order: "time",
    view_order_reversed: false,
    web_columns: ["tls", "icon", "path", "method", "status", "size", "time"],
    web_debug: false,
    web_host: "127.0.0.1",
    web_open_browser: true,
    web_port: 8081,
    web_static_viewer: "",
    websocket: true,
}
