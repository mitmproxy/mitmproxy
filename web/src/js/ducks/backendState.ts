/**
 * serverState houses properties about the current mitmproxy instance that are not options,
 * e.g. the list of available content views or the current version.
 */

export const RECEIVE = 'STATE_RECEIVE'
export const UPDATE = 'STATE_UPDATE'

export interface ServerInfo {
    description: string
    full_spec: string
    is_running: boolean
    last_exception: string | null
    listen_addrs: [string, number][]
    type: string
    wireguard_conf?: string
}

export interface BackendState {
    available: boolean
    version: string
    contentViews: string[]
    servers: ServerInfo[]
}

export const defaultState: BackendState = {
    available: false,
    version: "",
    contentViews: [],
    servers: [],
};

export default function reducer(state = defaultState, action): BackendState {
    switch (action.type) {
        case RECEIVE:
        case UPDATE:
            return {
                ...state,
                available: true,
                ...action.data
            }
        default:
            return state
    }
}
