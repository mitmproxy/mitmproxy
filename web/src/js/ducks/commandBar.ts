export const TOGGLE_VISIBILITY = "COMMANDBAR_TOGGLE_VISIBILITY";

interface CommandBarState {
    visible: boolean;
}

export const defaultState: CommandBarState = {
    visible: false,
};

export default function reducer(state = defaultState, action): CommandBarState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        default:
            return state;
    }
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}
