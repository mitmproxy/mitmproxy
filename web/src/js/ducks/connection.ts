export enum ConnectionState {
    INIT = "CONNECTION_INIT",
    FETCHING = "CONNECTION_FETCHING", // WebSocket is established, but still fetching resources.
    ESTABLISHED = "CONNECTION_ESTABLISHED",
    ERROR = "CONNECTION_ERROR",
    OFFLINE = "CONNECTION_OFFLINE", // indicates that there is no live (websocket) backend.
}

interface ConnState {
    state: ConnectionState;
    message?: string;
}

const defaultState: ConnState = {
    state: ConnectionState.INIT,
    message: undefined,
};

export default function reducer(state = defaultState, action): ConnState {
    switch (action.type) {
        case ConnectionState.ESTABLISHED:
        case ConnectionState.FETCHING:
        case ConnectionState.ERROR:
        case ConnectionState.OFFLINE:
            return {
                state: action.type,
                message: action.message,
            };

        default:
            return state;
    }
}

export function startFetching() {
    return { type: ConnectionState.FETCHING };
}

export function connectionEstablished() {
    return { type: ConnectionState.ESTABLISHED };
}

export function connectionError(message) {
    return { type: ConnectionState.ERROR, message };
}

export function setOffline() {
    return { type: ConnectionState.OFFLINE };
}
