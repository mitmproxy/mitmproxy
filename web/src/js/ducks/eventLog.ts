import * as store from "./utils/store";

export const ADD = "EVENTS_ADD";
export const RECEIVE = "EVENTS_RECEIVE";
export const TOGGLE_VISIBILITY = "EVENTS_TOGGLE_VISIBILITY";
export const TOGGLE_FILTER = "EVENTS_TOGGLE_FILTER";

type LogLevel = "debug" | "info" | "web" | "warn" | "error";

export interface EventLogItem extends store.Item {
    message: string;
    level: LogLevel;
}

interface EventLogState extends store.State<EventLogItem> {
    visible: boolean;
    filters: { [level in LogLevel]: boolean };
}

const defaultState: EventLogState = {
    visible: false,
    filters: { debug: false, info: true, web: true, warn: true, error: true },
    ...store.defaultState,
};

export default function reduce(
    state: EventLogState = defaultState,
    action
): EventLogState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        case TOGGLE_FILTER:
            const filters = {
                ...state.filters,
                [action.filter]: !state.filters[action.filter],
            };
            return {
                ...state,
                filters,
                ...store.reduce(
                    state,
                    store.setFilter<EventLogItem>((log) => filters[log.level])
                ),
            };

        case ADD:
        case RECEIVE:
            return {
                ...state,
                ...store.reduce(
                    state,
                    store[action.cmd](
                        action.data,
                        (log) => state.filters[log.level]
                    )
                ),
            };

        default:
            return state;
    }
}

export function toggleFilter(filter: LogLevel) {
    return { type: TOGGLE_FILTER, filter };
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}

export function add(message: string, level: LogLevel = "web") {
    let data = {
        id: Math.random().toString(),
        message,
        level,
    };
    return {
        type: ADD,
        cmd: "add",
        data,
    };
}
