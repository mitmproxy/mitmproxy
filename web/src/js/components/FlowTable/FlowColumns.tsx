import React, {ReactElement, useState} from 'react'
import {useDispatch} from 'react-redux'
import classnames from 'classnames'
import {canReplay, endTime, getTotalSize, RequestUtils, ResponseUtils, startTime} from '../../flow/utils'
import {formatSize, formatTimeDelta, formatTimeStamp} from '../../utils'
import * as flowActions from "../../ducks/flows";
import {Flow} from "../../flow";


type FlowColumnProps = {
    flow: Flow
}

interface FlowColumn {
    (props: FlowColumnProps): JSX.Element;

    headerName: string; // Shown in the UI
    sortKey: (flow: Flow) => any;
}

export const tls: FlowColumn = ({flow}) => {
    return (
        <td className={classnames('col-tls', flow.client_conn.tls_established ? 'col-tls-https' : 'col-tls-http')}/>
    )
}
tls.headerName = ''
tls.sortKey = flow => flow.type === "http" && flow.request.scheme

export const icon: FlowColumn = ({flow}) => {
    return (
        <td className="col-icon">
            <div className={classnames('resource-icon', getIcon(flow))}/>
        </td>
    )
}
icon.headerName = ''
icon.sortKey = flow => getIcon(flow)

const getIcon = (flow: Flow): string => {
    if (flow.type === "tcp" || flow.type === "dns") {
        return `resource-icon-${flow.type}`
    }
    if (flow.websocket) {
        return 'resource-icon-websocket'
    }
    if (!flow.response) {
        return 'resource-icon-plain'
    }

    var contentType = ResponseUtils.getContentType(flow.response) || ''

    if (flow.response.status_code === 304) {
        return 'resource-icon-not-modified'
    }
    if (300 <= flow.response.status_code && flow.response.status_code < 400) {
        return 'resource-icon-redirect'
    }
    if (contentType.indexOf('image') >= 0) {
        return 'resource-icon-image'
    }
    if (contentType.indexOf('javascript') >= 0) {
        return 'resource-icon-js'
    }
    if (contentType.indexOf('css') >= 0) {
        return 'resource-icon-css'
    }
    if (contentType.indexOf('html') >= 0) {
        return 'resource-icon-document'
    }

    return 'resource-icon-plain'
}

const mainPath = (flow: Flow): string => {
    switch (flow.type) {
        case "http":
            return RequestUtils.pretty_url(flow.request)
        case "tcp":
            return `${flow.client_conn.peername.join(':')} â†” ${flow.server_conn?.address?.join(':')}`
        case "dns":
            return `${flow.request.questions.map(q => `${q.name} ${q.type}`).join(", ")} = ${(flow.response?.answers.map(q => q.data).join(", ") ?? "...") || "?"}`
    }
}

export const path: FlowColumn = ({flow}) => {
    let err;
    if (flow.error) {
        if (flow.error.msg === "Connection killed.") {
            err = <i className="fa fa-fw fa-times pull-right"/>
        } else {
            err = <i className="fa fa-fw fa-exclamation pull-right"/>
        }
    }
    return (
        <td className="col-path">
            {flow.is_replay === "request" && (
                <i className="fa fa-fw fa-repeat pull-right"/>
            )}
            {flow.intercepted && (
                <i className="fa fa-fw fa-pause pull-right"/>
            )}
            {err}
            <span className="marker pull-right">{flow.marked}</span>
            {mainPath(flow)}
        </td>
    )
};
path.headerName = 'Path'
path.sortKey = flow => mainPath(flow)

export const method: FlowColumn = ({flow}) => <td className="col-method">{method.sortKey(flow)}</td>
method.headerName = 'Method'
method.sortKey = flow => {
    switch (flow.type) {
        case "http": return flow.websocket ? (flow.client_conn.tls_established ? "WSS" : "WS") : flow.request.method
        case "dns": return flow.request.op_code
        default: return flow.type.toUpperCase()
    }
}

export const status: FlowColumn = ({flow}) => {
    let color = 'darkred'

    if ((flow.type !== "http" && flow.type != "dns") || !flow.response)
        return <td className="col-status"/>

    if (100 <= flow.response.status_code && flow.response.status_code < 200) {
        color = 'green'
    } else if (200 <= flow.response.status_code && flow.response.status_code < 300) {
        color = 'darkgreen'
    } else if (300 <= flow.response.status_code && flow.response.status_code < 400) {
        color = 'lightblue'
    } else if (400 <= flow.response.status_code && flow.response.status_code < 500) {
        color = 'red'
    } else if (500 <= flow.response.status_code && flow.response.status_code < 600) {
        color = 'red'
    }

    return (
        <td className="col-status" style={{color: color}}>{status.sortKey(flow)}</td>
    )
}
status.headerName = 'Status'
status.sortKey = flow => {
    switch (flow.type) {
        case "http": return flow.response?.status_code
        case "dns": return flow.response?.response_code
        default: return undefined
    }
}

export const size: FlowColumn = ({flow}) => {
    return (
        <td className="col-size">{formatSize(getTotalSize(flow))}</td>
    )
};
size.headerName = 'Size'
size.sortKey = flow => getTotalSize(flow)


export const time: FlowColumn = ({flow}) => {
    const start = startTime(flow), end = endTime(flow);
    return (
        <td className="col-time">
            {start && end ? (
                formatTimeDelta(1000 * (end - start))
            ) : (
                '...'
            )}
        </td>
    )
}
time.headerName = 'Time'
time.sortKey = flow => {
    const start = startTime(flow), end = endTime(flow);
    return start && end && end - start;
}

export const timestamp: FlowColumn = ({flow}) => {
    const start = startTime(flow);
    return (
        <td className="col-timestamp">
            {start ? (
                formatTimeStamp(start)
            ) : (
                '...'
            )}
        </td>
    )
}
timestamp.headerName = 'Start time'
timestamp.sortKey = flow => startTime(flow)

const markers = {
    ":red_circle:": "ðŸ”´",
    ":orange_circle:": "ðŸŸ ",
    ":yellow_circle:": "ðŸŸ¡",
    ":green_circle:": "ðŸŸ¢",
    ":large_blue_circle:": "ðŸ”µ",
    ":purple_circle:": "ðŸŸ£",
    ":brown_circle:": "ðŸŸ¤",
}

export const quickactions: FlowColumn = ({flow}) => {
    const dispatch = useDispatch()
    let [open, setOpen] = useState(false)

    let resume_or_replay: ReactElement | null = null;
    if (flow.intercepted) {
        resume_or_replay = <a href="#" className="quickaction" onClick={() => dispatch(flowActions.resume(flow))}>
            <i className="fa fa-fw fa-play text-success"/>
        </a>;
    } else if (canReplay(flow)) {
        resume_or_replay = <a href="#" className="quickaction" onClick={() => dispatch(flowActions.replay(flow))}>
            <i className="fa fa-fw fa-repeat text-primary"/>
        </a>;
    }

    return (
        <td className={classnames("col-quickactions", {hover: open})} onClick={() => 0}>
            <div>
                {resume_or_replay}
            </div>
        </td>
    )
}

quickactions.headerName = ''
quickactions.sortKey = flow => 0;

export default {
    icon,
    method,
    path,
    quickactions,
    size,
    status,
    time,
    timestamp,
    tls
};
