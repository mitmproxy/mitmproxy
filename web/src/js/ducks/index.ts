import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
