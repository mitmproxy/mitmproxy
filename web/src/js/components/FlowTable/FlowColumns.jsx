import React from 'react'
import {useDispatch} from 'react-redux'
import classnames from 'classnames'
import {RequestUtils, ResponseUtils} from '../../flow/utils.js'
import {formatSize, formatTimeDelta, formatTimeStamp} from '../../utils.js'
import * as flowActions from "../../ducks/flows";

export const defaultColumnNames = ["tls", "icon", "path", "method", "status", "size", "time"]

export function TLSColumn({flow}) {
    return (
        <td className={classnames('col-tls', flow.request.scheme === 'https' ? 'col-tls-https' : 'col-tls-http')}/>
    )
}

TLSColumn.headerClass = 'col-tls'
TLSColumn.headerName = ''

export function IconColumn({flow}) {
    return (
        <td className="col-icon">
            <div className={classnames('resource-icon', IconColumn.getIcon(flow))}/>
        </td>
    )
}

IconColumn.headerClass = 'col-icon'
IconColumn.headerName = ''

IconColumn.getIcon = flow => {
    if (!flow.response) {
        return 'resource-icon-plain'
    }

    var contentType = ResponseUtils.getContentType(flow.response) || ''

    // @todo We should assign a type to the flow somewhere else.
    if (flow.response.status_code === 304) {
        return 'resource-icon-not-modified'
    }
    if (300 <= flow.response.status_code && flow.response.status_code < 400) {
        return 'resource-icon-redirect'
    }
    if (contentType.indexOf('image') >= 0) {
        return 'resource-icon-image'
    }
    if (contentType.indexOf('javascript') >= 0) {
        return 'resource-icon-js'
    }
    if (contentType.indexOf('css') >= 0) {
        return 'resource-icon-css'
    }
    if (contentType.indexOf('html') >= 0) {
        return 'resource-icon-document'
    }

    return 'resource-icon-plain'
}

export function PathColumn({flow}) {

    let err;
    if (flow.error) {
        if (flow.error.msg === "Connection killed.") {
            err = <i className="fa fa-fw fa-times pull-right"/>
        } else {
            err = <i className="fa fa-fw fa-exclamation pull-right"/>
        }
    }
    return (
        <td className="col-path">
            {flow.request.is_replay && (
                <i className="fa fa-fw fa-repeat pull-right"/>
            )}
            {flow.intercepted && (
                <i className="fa fa-fw fa-pause pull-right"/>
            )}
            {err}
            {RequestUtils.pretty_url(flow.request)}
        </td>
    )
}

PathColumn.headerClass = 'col-path'
PathColumn.headerName = 'Path'

export function MethodColumn({flow}) {
    return (
        <td className="col-method">{flow.request.method}</td>
    )
}

MethodColumn.headerClass = 'col-method'
MethodColumn.headerName = 'Method'

export function StatusColumn({flow}) {
    let color = 'darkred';

    if (flow.response && 100 <= flow.response.status_code && flow.response.status_code < 200) {
        color = 'green'
    } else if (flow.response && 200 <= flow.response.status_code && flow.response.status_code < 300) {
        color = 'darkgreen'
    } else if (flow.response && 300 <= flow.response.status_code && flow.response.status_code < 400) {
        color = 'lightblue'
    } else if (flow.response && 400 <= flow.response.status_code && flow.response.status_code < 500) {
        color = 'lightred'
    } else if (flow.response && 500 <= flow.response.status_code && flow.response.status_code < 600) {
        color = 'lightred'
    }

    return (
        <td className="col-status" style={{color: color}}>{flow.response && flow.response.status_code}</td>
    )
}

StatusColumn.headerClass = 'col-status'
StatusColumn.headerName = 'Status'

export function SizeColumn({flow}) {
    return (
        <td className="col-size">{formatSize(SizeColumn.getTotalSize(flow))}</td>
    )
}

SizeColumn.getTotalSize = flow => {
    let total = flow.request.contentLength
    if (flow.response) {
        total += flow.response.contentLength || 0
    }
    return total
}

SizeColumn.headerClass = 'col-size'
SizeColumn.headerName = 'Size'

export function TimeColumn({flow}) {
    return (
        <td className="col-time">
            {flow.response ? (
                formatTimeDelta(1000 * (flow.response.timestamp_end - flow.request.timestamp_start))
            ) : (
                '...'
            )}
        </td>
    )
}

TimeColumn.headerClass = 'col-time'
TimeColumn.headerName = 'Time'

export function TimeStampColumn({flow}) {
    return (
        <td className="col-start">
            {flow.request.timestamp_start ? (
                formatTimeStamp(flow.request.timestamp_start)
            ) : (
                '...'
            )}
        </td>
    )
}

TimeStampColumn.headerClass = 'col-timestamp'
TimeStampColumn.headerName = 'TimeStamp'

export function QuickActionsColumn({flow}) {
    const dispatch = useDispatch()

    function resume(e) {
        dispatch(flowActions.resume(flow))
        e.preventDefault()
        e.stopPropagation()
    }

    return (
        <td className="col-quickactions">
            <div>
                <div className="quickaction"><i className="fa fa-fw fa-ellipsis-h"/></div>
                {flow.intercepted
                    ? <div className="quickaction" onClick={resume}><i className="fa fa-fw fa-play text-success"/></div>
                    : null}
            </div>
        </td>
    )
}

QuickActionsColumn.headerClass = 'col-quickactions'
QuickActionsColumn.headerName = ''


export const columns = {};
for (let col of [
    TLSColumn,
    IconColumn,
    PathColumn,
    MethodColumn,
    StatusColumn,
    TimeStampColumn,
    SizeColumn,
    TimeColumn,
    QuickActionsColumn,
]) {
    columns[col.name.replace(/Column$/, "").toLowerCase()] = col;
}
