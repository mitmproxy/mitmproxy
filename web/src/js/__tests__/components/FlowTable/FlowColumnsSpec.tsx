import * as React from "react"
import renderer from 'react-test-renderer'
import FlowColumns from '../../../components/FlowTable/FlowColumns'
import {TFlow, TTCPFlow} from '../../ducks/tutils'
import {render} from "../../test-utils";

test("should render columns", async () => {
    const tflow = TFlow();
    Object.entries(FlowColumns).forEach(([name, Col]) => {
        const {asFragment} = render(<table>
            <tbody>
            <tr><Col flow={tflow}/></tr>
            </tbody>
        </table>)
        expect(asFragment()).toMatchSnapshot(name);
    })
});


describe('Flowcolumns Components', () => {
    it('should render IconColumn', () => {
        let tcpflow = TTCPFlow(),
            iconColumn = renderer.create(<FlowColumns.icon flow={tcpflow}/>),
            tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()

        let tflow = {...TFlow(), websocket: undefined};
        iconColumn = renderer.create(<FlowColumns.icon flow={tflow}/>)
        tree = iconColumn.toJSON()
        // plain
        expect(tree).toMatchSnapshot()
        // not modified
        tflow.response.status_code = 304
        iconColumn = renderer.create(<FlowColumns.icon flow={tflow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // redirect
        tflow.response.status_code = 302
        iconColumn = renderer.create(<FlowColumns.icon flow={tflow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // image
        let imageFlow = {...TFlow(), websocket: undefined}
        imageFlow.response.headers = [['Content-Type', 'image/jpeg']]
        iconColumn = renderer.create(<FlowColumns.icon flow={imageFlow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // javascript
        let jsFlow = {...TFlow(), websocket: undefined}
        jsFlow.response.headers = [['Content-Type', 'application/x-javascript']]
        iconColumn = renderer.create(<FlowColumns.icon flow={jsFlow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // css
        let cssFlow = {...TFlow(), websocket: undefined}
        cssFlow.response.headers = [['Content-Type', 'text/css']]
        iconColumn = renderer.create(<FlowColumns.icon flow={cssFlow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // html
        let htmlFlow = {...TFlow(), websocket: undefined}
        htmlFlow.response.headers = [['Content-Type', 'text/html']]
        iconColumn = renderer.create(<FlowColumns.icon flow={htmlFlow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // default
        let fooFlow = {...TFlow(), websocket: undefined}
        fooFlow.response.headers = [['Content-Type', 'foo']]
        iconColumn = renderer.create(<FlowColumns.icon flow={fooFlow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
        // no response
        let noResponseFlow = {...TFlow(), response: undefined}
        iconColumn = renderer.create(<FlowColumns.icon flow={noResponseFlow}/>)
        tree = iconColumn.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should render pathColumn', () => {
        let tflow = TFlow(),
            pathColumn = renderer.create(<FlowColumns.path flow={tflow}/>),
            tree = pathColumn.toJSON()
        expect(tree).toMatchSnapshot()

        tflow.error.msg = 'Connection killed.'
        tflow.intercepted = true
        pathColumn = renderer.create(<FlowColumns.path flow={tflow}/>)
        tree = pathColumn.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should render TimeColumn', () => {
        let tflow = TFlow(),
            timeColumn = renderer.create(<FlowColumns.time flow={tflow}/>),
            tree = timeColumn.toJSON()
        expect(tree).toMatchSnapshot()

        let noResponseFlow = {...tflow, response: undefined}
        timeColumn = renderer.create(<FlowColumns.time flow={noResponseFlow}/>)
        tree = timeColumn.toJSON()
        expect(tree).toMatchSnapshot()
    })
})
