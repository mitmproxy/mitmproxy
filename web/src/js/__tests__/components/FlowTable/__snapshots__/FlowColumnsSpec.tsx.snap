// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flowcolumns Components should render IconColumn 1`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-tcp"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 2`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-plain"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 3`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-not-modified"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 4`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-redirect"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 5`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-image"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 6`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-js"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 7`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-css"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 8`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-document"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 9`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-plain"
  />
</td>
`;

exports[`Flowcolumns Components should render IconColumn 10`] = `
<td
  className="col-icon"
>
  <div
    className="resource-icon resource-icon-websocket"
  />
</td>
`;

exports[`Flowcolumns Components should render TimeColumn 1`] = `
<td
  className="col-time"
>
  5s
</td>
`;

exports[`Flowcolumns Components should render TimeColumn 2`] = `
<td
  className="col-time"
>
  5s
</td>
`;

exports[`Flowcolumns Components should render pathColumn 1`] = `
<td
  className="col-path"
>
  <i
    className="fa fa-fw fa-exclamation pull-right"
  />
  <span
    className="marker pull-right"
  >
    
  </span>
  http://address:22/path
</td>
`;

exports[`Flowcolumns Components should render pathColumn 2`] = `
<td
  className="col-path"
>
  <i
    className="fa fa-fw fa-pause pull-right"
  />
  <i
    className="fa fa-fw fa-times pull-right"
  />
  <span
    className="marker pull-right"
  >
    
  </span>
  http://address:22/path
</td>
`;

exports[`should render columns: icon 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-icon"
        >
          <div
            class="resource-icon resource-icon-websocket"
          />
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: method 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-method"
        >
          WSS
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: path 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-path"
        >
          <i
            class="fa fa-fw fa-exclamation pull-right"
          />
          <span
            class="marker pull-right"
          />
          http://address:22/path
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: quickactions 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-quickactions"
        >
          <div />
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: size 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-size"
        >
          43b
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: status 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-status"
          style="color: darkgreen;"
        >
          200
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: time 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-time"
        >
          5s
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: timestamp 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-timestamp"
        >
          1999-12-31 23:00:00.000
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should render columns: tls 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-tls col-tls-https"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
